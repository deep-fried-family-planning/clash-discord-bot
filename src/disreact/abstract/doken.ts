import {NONE_INT, NONE_STR, Rest} from '#src/disreact/abstract/index.ts';

const TWO_SECONDS_MS      = 2 * 1000;
const FOURTEEN_MINUTES_MS = 14 * 60 * 1000;
const TWO_MINUTES_MS      = 2 * 60 * 1000;

export type T = {
  app  : string;
  id   : string;
  token: string;
  ttl  : number;
  type : Rest.Tx;
  flags: number;
};

export type TEncoded = {
  id   : string;
  token: string;
  ttl  : string;
  type : string;
  flags: string;
};

export const initialTTL = () => Date.now() + TWO_SECONDS_MS;
export const deferTTL   = () => Date.now() + FOURTEEN_MINUTES_MS;

export const makeEmpty = (): T => ({
  app  : '',
  id   : NONE_STR,
  token: NONE_STR,
  ttl  : 0,
  type : Rest.DEFER_SOURCE,
  flags: 0,
});

export const makeFromRest = (rest: Rest.Interaction): T => ({
  app  : rest.application_id,
  id   : rest.id,
  token: rest.token,
  type : Rest.DEFER_SOURCE,
  ttl  : initialTTL(),
  flags: 0,
});

export const validateTTL = (doken: T): T | null => {
  if (doken.id === NONE_STR) {
    return null;
  }
  if (doken.ttl === NONE_INT) {
    return null;
  }
  if ((doken.ttl - TWO_MINUTES_MS) > Date.now()) {
    return doken;
  }
  return null;
};

export const encode = (doken?: T): TEncoded => doken
  ? {
    id   : doken.id,
    token: doken.token,
    ttl  : `${doken.ttl}`,
    type : `${doken.type}`,
    flags: `${doken.flags}`,
  }
  : {
    id   : NONE_STR,
    token: NONE_STR,
    ttl  : '0',
    type : '0',
    flags: '0',
  };
