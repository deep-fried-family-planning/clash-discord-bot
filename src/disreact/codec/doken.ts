import {UtcNow} from '#src/disreact/codec/rest/shared.ts'
import {RDT} from '#src/disreact/codec/re-exports.ts'
import {DokenMemory} from '#src/disreact/runtime/DokenMemory.ts'
import {DR, DT, E, pipe, S} from '#src/internal/pure/effect.ts'
import {DateTime, ParseResult} from 'effect'
import {DateTimeUtcFromNumber, DateTimeUtcFromSelf, RedactedFromSelf} from 'effect/Schema'
import {CallbackType} from 'src/disreact/codec/rest/callback-type.ts'
import {Flag} from 'src/disreact/codec/rest/flag.ts'
import { DAPIIX} from 'src/disreact/codec/rest/dapi-ix.ts'


export type Fresh = typeof Fresh.T.Type

export namespace Fresh {
  export const TAG = 'Fresh' as const

  export const T = S.Struct({
    app : S.String,
    id  : S.String,
    val : RedactedFromSelf(S.String),
    type: CallbackType.All,
    flag: Flag.All,
    ttl : UtcNow,
  }).pipe(S.attachPropertySignature('_tag', TAG))

  export const FromRequest = S.transformOrFail(DAPIIX.BaseBody, T, {
    strict: true,
    decode: (request) =>
      pipe(
        DateTime.now,
        E.map((now) =>
          ({
            app : request.application_id,
            id  : request.id,
            type: CallbackType.FRESH,
            flag: Flag.FRESH,
            val : request.token,
            ttl : DateTime.add(now, {millis: 2000}),
          } as const),
        ),
      ),
    encode: () => {throw new Error()},
  })
}

export type Defer = typeof Defer.T.Type

export namespace Defer {
  export type Maybe = typeof Maybe.Type

  export const TAG = 'Defer' as const

  export const T = S.Struct({
    _tag: S.tag(TAG),
    id  : S.String,
    val : RedactedFromSelf(S.String),
    type: CallbackType.Defer,
    flag: Flag.Defined,
    ttl : DateTimeUtcFromSelf,
  })

  export const A = S.TemplateLiteralParser(
    'd/', CallbackType.Defer,
    '/', Flag.Defined,
    '/', S.String,
    '/', DateTimeUtcFromNumber,
    '/', S.Redacted(S.String),
  )

  export const D = S.transform(A, T, {
    strict: true,
    decode: ([, type, , flag, , id, , ttl, , val]) =>
      ({
        _tag: TAG,
        id,
        type,
        flag,
        ttl,
        val,
      }),
    encode: ({id, type, flag, ttl, val}) =>
      [
        'd/', type, '/', flag, '/', id, '/', ttl, '/', val,
      ] as const,
  })

  export const Maybe = S.UndefinedOr(T)
}


export type Cache = typeof Cache.T.Type

export namespace Cache {
  export const TAG = 'Cache' as const

  export const T = S.Struct({
    _tag: S.tag(TAG),
    id  : S.String,
    type: CallbackType.Defer,
    flag: Flag.Defined,
    ttl : DateTimeUtcFromSelf,
  })

  export const A = S.TemplateLiteralParser(
    'c/', CallbackType.Defer,
    '/', Flag.Defined,
    '/', S.String,
    '/', DateTimeUtcFromNumber,
  )

  export const D = S.transform(A, T, {
    strict: true,
    decode: ([, type, , flag, , id, , ttl]) =>
      ({
        _tag: TAG,
        id,
        type,
        flag,
        ttl,
      }),
    encode: ({id, type, flag, ttl}) =>
      [
        'c/', type, '/', flag, '/', id, '/', ttl,
      ] as const,
  })

  // export const FromDefer = S.transformOrFail(Defer.T, T, {
  //   strict: true,
  //   decode: (defer) =>
  //     pipe(
  //       DokenMemory.save(defer),
  //       E.catchAll(() => E.fail(new ParseResult.Unexpected(undefined))),
  //       E.map(() =>
  //         ({
  //           _tag: TAG,
  //           id  : defer.id,
  //           type: defer.type,
  //           flag: defer.flag,
  //           ttl : defer.ttl,
  //         }),
  //       ),
  //     ),
  //   encode: () => {throw new Error()},
  // })
}


export type Spent = typeof Spent.T.Type

export namespace Spent {
  export type Spent = typeof T.Type
  export const TAG = 'Spent' as const

  export const T = S.Struct({
    _tag: S.tag(TAG),
    id  : S.String,
    type: CallbackType.Spent,
    flag: Flag.Defined,
    val : RedactedFromSelf(S.String),
  })

  export const A = S.TemplateLiteralParser(
    's/', CallbackType.Spent,
    '/', Flag.Defined,
    '/', S.String,
  )

  export const D = S.transform(A, T, {
    strict: true,
    decode: ([, type, , flag, , id]) =>
      ({
        _tag: TAG,
        id,
        type,
        flag,
        val : RDT.make('-'),
      }),
    encode: ({id, type, flag}) =>
      [
        's/', type, '/', flag, '/', id,
      ] as const,
  })

  export const FromFresh = S.transform(Fresh.T, T, {
    strict: true,
    decode: (fresh) =>
      ({
        _tag: TAG,
        id  : fresh.id,
        type: fresh.type as any,
        flag: fresh.flag as any,
        val : fresh.val,
      }),
    encode: () => {throw new Error()},
  })
}


export * as Doken from '#src/disreact/codec/doken.ts'
export type Doken = typeof T.Type

export const T = S.Union(
  Defer.T,
  Cache.T,
  Spent.T,
  // Fresh.T.pipe(S.attachPropertySignature('_tag', Fresh.TAG)),
)

export const A = S.Union(
  Defer.A,
  Cache.A,
  Spent.A,
)

export const D = S.Union(
  Defer.D,
  Cache.D,
  Spent.D,
)

export const makeFreshFromRequest = (request: DAPIIX.Body, startMs?: number): E.Effect<Fresh> =>
  pipe(
    startMs
      ? E.succeed(DT.unsafeMake(startMs))
      : DT.now,
    E.map((now) =>
      ({
        _tag: 'Fresh',
        app : request.application_id,
        id  : request.id,
        type: CallbackType.FRESH,
        flag: Flag.FRESH,
        val : RDT.isRedacted(request.token) ? request.token : RDT.make(request.token),
        ttl : DateTime.add(now, {millis: 2000}),
      }),
    ),
  )

export const makeOptimizedDeferFromFresh = (request: DAPIIX.Body, fresh: Fresh) =>
  pipe(
    DT.isPast(fresh.ttl),
    E.andThen((isPast) => {
      if (isPast) {
        return E.fail(new ParseResult.Unexpected('Expired'))
      }
      return E.succeed({
        _tag: 'Defer',
        id  : fresh.id,
        type: CallbackType.UPDATE_DEFER,
        flag: request.message?.flags === 64 ? Flag.PRIVATE : Flag.PUBLIC,
        ttl : DT.addDuration(fresh.ttl, DR.minutes(12)),
        val : fresh.val,
      } satisfies Defer)
    }),
  )

export const makeDeferFromFresh = (request: DAPIIX.Body, fresh: Fresh, flags?: number) =>
  pipe(
    DT.isPast(fresh.ttl),
    E.andThen((isPast) => {
      if (isPast) {
        return E.fail(new ParseResult.Unexpected('Expired'))
      }
      const msgFlags = request.message?.flags === 64 ? 2 : 1
      return E.succeed({
        _tag: 'Defer',
        id  : fresh.id,
        type: flags === msgFlags ? CallbackType.UPDATE_DEFER : CallbackType.SOURCE_DEFER,
        flag: flags as any ?? 2,
        ttl : DT.addDuration(fresh.ttl, DR.minutes(12)),
        val : RDT.isRedacted(fresh.val) ? fresh.val : RDT.make(fresh.val),
      } satisfies Defer)
    }),
  )
