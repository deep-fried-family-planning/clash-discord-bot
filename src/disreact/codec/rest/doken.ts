import {DokenMemory} from '#src/disreact/utils/DokenMemory.ts';
import {D, DT, E, hole} from '#src/disreact/utils/re-exports.ts';
import {DR, pipe, S} from '#src/internal/pure/effect.ts';
import {Duration, Either, Option} from 'effect';
import {DateTime} from 'effect';
import { Redacted} from 'effect';
import {Snowflake} from 'src/disreact/codec/snowflake.ts';

export * as Doken from '#src/disreact/codec/rest/doken.ts';
export type Doken =
  | Fresh
  | Active
  | Cached
  | Single
  | Modal
  | Source
  | Update
  | Never;

export const isFresh = (doken: Doken): doken is Fresh => doken._tag === FRESH;
export const isActive = (doken: Doken): doken is Active => doken._tag === ACTIVE;
export const isCached = (doken: Doken): doken is Cached => doken._tag === CACHED;
export const isSingle = (doken: Doken): doken is Single => doken._tag === SINGLE;
export const isModal = (doken: Doken): doken is Modal => doken._tag === MODAL;
export const isSource = (doken: Doken): doken is Source => doken._tag === SOURCE;
export const isUpdate = (doken: Doken): doken is Update => doken._tag === UPDATE;
export const isNever = (doken: Doken): doken is Never => doken._tag === NEVER;

export type Value = Exclude<Doken, Cached>;

export type Create =
  | Modal
  | Source
  | Update;

export type Serial =
  | Single
  | Active
  | Cached;

export const value = (doken: Doken) => Redacted.value(doken.val);

const Base = S.Struct({
  id : Snowflake.Id,
  ttl: S.DateTimeUtcFromSelf,
  val: S.RedactedFromSelf(S.String),
  app: Snowflake.Id,
  eph: S.optional(S.Boolean),
});

export const FRESH_OFFSET = DR.seconds(2);
export const ACTIVE_OFFSET = DR.minutes(12);

export const FRESH = 'Fresh';
export type Fresh = S.Schema.Type<typeof Fresh>;
export const Fresh = pipe(
  S.Struct({
    id            : Snowflake.Id,
    token         : S.RedactedFromSelf(S.String),
    application_id: Snowflake.Id,
    message       : S.optional(S.Struct({flags: S.optional(S.Number)})),
  }),
  S.transform(
    S.Struct({
      ...Base.fields,
      ttl: Snowflake.TimeToLive(FRESH_OFFSET),
      val: S.RedactedFromSelf(S.String),
    }),
    {
      encode: hole,
      decode: (req) =>
        ({
          id : req.id,
          ttl: req.id,
          val: req.token,
          app: req.application_id,
          eph: req.message?.flags === 64,
        }),
    },
  ),
  S.attachPropertySignature('_tag', FRESH),
  S.mutable,
);

export const NEVER = 'Never';
export type Never = typeof Never.Type;
export const Never = pipe(
  Base,
  S.attachPropertySignature('_tag', NEVER),
  S.mutable,
);

export const never = (): Never =>
  ({
    _tag: NEVER,
  } as unknown as Never);

export const MODAL = 'Modal';
export type Modal = typeof Modal.Type;
export const Modal = pipe(
  Base,
  S.attachPropertySignature('_tag', MODAL),
  S.mutable,
);

export const modal = (fresh: Fresh): Modal =>
  ({
    ...fresh,
    _tag: MODAL,
  });

export const SOURCE = 'Source';
export type Source = typeof Source.Type;
export const Source = pipe(
  Base,
  S.attachPropertySignature('_tag', SOURCE),
  S.mutable,
);

export const source = (fresh: Fresh): Source =>
  ({
    ...fresh,
    eph : !fresh.eph,
    _tag: SOURCE,
  });

export const UPDATE = 'Update';
export type Update = typeof Update.Type;
export const Update = pipe(
  Base,
  S.attachPropertySignature('_tag', UPDATE),
  S.mutable,
);

export const update = (fresh: Fresh): Update =>
  ({
    ...fresh,
    _tag: UPDATE,
  });

export const SINGLE = 'Single' as const;
export type Single = typeof Single.Type;
export const Single = pipe(
  S.TemplateLiteralParser(
    's',
  ),
  S.transform(
    Base,
    {
      encode: () =>
        [
          's',
        ] as const,
      decode: () =>
        ({
          id : '',
          ttl: DT.unsafeMake(0),
          val: Redacted.make(''),
          app: '',
        }),
    },
  ),
  S.attachPropertySignature('_tag', SINGLE),
  S.mutable,
);

export const single = (doken: Doken): Single =>
  ({
    ...doken,
    _tag: SINGLE,
  });

export const synthetic = (): Single =>
  ({
    _tag: SINGLE,
    id  : 'synthetic',
    ttl : DT.unsafeMake(0),
    val : Redacted.make(''),
    app : '',
  });

export const ACTIVE = 'Active';
export type Active = typeof Active.Type;
export const Active = pipe(
  S.TemplateLiteralParser(
    'a/', Snowflake.Id,
    '/', S.Redacted(S.String),
  ),
  S.transform(
    S.Struct({
      ...Base.fields,
      ttl: Snowflake.TimeToLive(ACTIVE_OFFSET),
    }),
    {
      encode: ({id, val}) =>
        [
          'a/', id, '/', val,
        ] as const,
      decode: ([, id, , val]) =>
        ({
          id : id,
          ttl: id,
          val: val,
          app: '',
        }),
    },
  ),
  S.attachPropertySignature('_tag', ACTIVE),
  S.mutable,
);

export const active = (doken: Doken): Active => {
  if (doken._tag === 'Fresh') {
    return {
      ...doken,
      _tag: ACTIVE,
      ttl : DateTime.addDuration(doken.ttl, ACTIVE_OFFSET),
    };
  }

  return {
    ...doken,
    _tag: ACTIVE,
  };
};

export const CACHED = 'Cached';
export type Cached = typeof Cached.Type;
export const Cached = pipe(
  S.TemplateLiteralParser(
    'c/', Snowflake.Id,
  ),
  S.transform(
    S.Struct({
      ...Base.fields,
      ttl: Snowflake.TimeToLive(ACTIVE_OFFSET),
    }),
    {
      encode: ({id}) =>
        [
          'c/', id,
        ] as const,
      decode: ([, id]) =>
        ({
          id : id,
          ttl: id,
          val: Redacted.make(''),
          app: '',
        }),
    },
  ),
  S.attachPropertySignature('_tag', CACHED),
  S.mutable,
);

export const cache = (doken: Doken): Cached =>
  ({
    ...doken,
    _tag: CACHED,
  });

export const Serial = S.Union(
  Active,
  Cached,
  Single,
);

export const convertSerial = (doken: Doken): Serial => {
  switch (doken._tag) {
    case ACTIVE:
      return doken;
    case SINGLE:
      return doken;
    case CACHED:
      return doken;
    default:
      return single(doken);
  }
};

export const convertCached = (doken: Serial): Serial => {
  switch (doken._tag) {
    case ACTIVE:
      return cache(doken);
    default:
      return doken;
  }
};

export class DokenDefect extends D.TaggedError('disreact/DokenDefect')<{
  msg?: string;
}> {}

export const ttlEither = (self: Doken | undefined, now: DateTime.Utc) =>
  pipe(
    self,
    Either.fromNullable(() => undefined),
    Either.flatMap((doken) =>
      DateTime.distanceDurationEither(now, doken.ttl),
    ),
    Either.map((delay) => [delay, self!] as const),
    Either.mapLeft(() => undefined),
  );

export const reduce = (state: Doken, action: Doken) => {
  if (action._tag === NEVER) {
    return action;
  }
  if (
    state._tag === NEVER ||
    state._tag === ACTIVE ||
    state._tag === MODAL
  ) {
    return state;
  }
  if (
    state._tag === FRESH
  ) {
    return action;
  }
};

export const resolveSerial = (fresh: Doken.Fresh, serial?: Doken.Serial) => {
  if (!serial || Doken.isSingle(serial)) {
    return E.succeed(undefined);
  }
  if (Doken.isActive(serial)) {
    return pipe(
      DateTime.isFuture(serial.ttl),
      E.map((isFuture) => isFuture ? serial : undefined),
    );
  }
  return pipe(
    DokenMemory.use((memory) => memory.load(serial.id)),
    E.orElseSucceed(() => undefined),
    E.timeoutTo({
      duration : Duration.seconds(1),
      onTimeout: () => undefined,
      onSuccess: (cached) => {
        if (!cached) {
          return cached;
        }
        cached.app = fresh.app;
        return cached;
      },
    }),
    E.whenEffect(DateTime.isFuture(serial.ttl)),
    E.map(Option.getOrUndefined),
  );
};
