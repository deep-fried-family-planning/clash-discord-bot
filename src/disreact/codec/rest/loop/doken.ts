import {NONE} from '#src/disreact/codec/common/index.ts';
import {DT, RDT, S} from '#src/internal/pure/effect.ts';
import type * as Route from '#src/disreact/codec/rest/route/index.ts';



const DT_FRESH_TOTAL_MS = 3000; // 3 seconds
const DT_FRESH_OFFSET_MS = 1000; // 1 second
const DT_FRESH_SAFETY_MS = DT_FRESH_TOTAL_MS - DT_FRESH_OFFSET_MS;
const DT_DEFER_TOTAL_MS = 900000; // 15 minutes
const DT_DEFER_OFFSET_MS = 120000; // 2 minutes
const DT_DEFER_SAFETY_MS = DT_DEFER_TOTAL_MS - DT_DEFER_OFFSET_MS;

const EXPIRED = -1;
const FRESH = 0;
const SPENT_SOURCE = 4;
const SPENT_UPDATE = 1;
const SPENT_DIALOG = 9;
const DEFER_SOURCE = 5;
const DEFER_UPDATE = 6;

const VALUE_PARTIAL = 'P';
const VALUE_NONE = 'N';

const NOT_SET = -1;
const PUBLIC = 0;
export const EPHEMERAL = 1;



export const T = S.Struct({
  id       : S.String,
  ttl      : S.DateTimeUtc,
  ephemeral: S.Int,
  type     : S.Int,
  value    : S.RedactedFromSelf(S.String),
});

export type T = S.Schema.Type<typeof T>;

export const isFresh = (doken: T): boolean => doken.type === FRESH;
export const isSpent = (doken: T): boolean => RDT.value(doken.value) === VALUE_NONE || doken.type === SPENT_SOURCE || doken.type === SPENT_UPDATE || doken.type === SPENT_DIALOG;
export const isDefer = (doken: T): boolean => doken.type === DEFER_SOURCE || doken.type === DEFER_UPDATE;
export const isPublic = (doken: T): boolean => doken.ephemeral === PUBLIC;
export const isEphemeral = (doken: T): boolean => doken.ephemeral === EPHEMERAL || doken.ephemeral === NOT_SET;
export const isPartial = (doken: T): boolean => !isSpent(doken) && RDT.value(doken.value) === VALUE_PARTIAL;
export const isExpired = (doken: T): boolean => DT.lessThan(doken.ttl, DT.unsafeNow());

export const makeFresh = (request: any): T => {
  return {
    id       : request.id,
    ttl      : DT.unsafeMake(Date.now() + DT_FRESH_SAFETY_MS),
    ephemeral: NOT_SET,
    component: FRESH,
    value    : RDT.make(request.token),
  };
};

export const deferFresh = (
  doken: T,
  ephemeral: number,
  type: number,
  time = DT.unsafeNow(),
): T => {
  return {
    ...doken,
    ttl      : DT.add(time, {millis: DT_DEFER_SAFETY_MS}),
    ephemeral: ephemeral,
    component: type,
  };
};

export const makeFromMessageRoute = (route: Route.Message.T): T | undefined => {
  if (
    !route.id
    || route.id === NONE
    || !route.token
    || !route.type
    || !route.ephemeral
    || !route.ttl
    || DT.unsafeIsPast(route.ttl)
  ) {
    return undefined;
  }

  return {
    id       : route.id,
    ttl      : route.ttl,
    ephemeral: route.ephemeral,
    component: route.type,
    value    : route.token,
  };
};

export const validateDecode = (doken: T): T | undefined => {
  if (isExpired(doken)) {
    return undefined;
  }
  return doken;
};

export const validateEncode = (doken: T) => {
  if (isExpired(doken)) {
    throw new Error('Token is expired.');
  }

  if (isPublic(doken)) {
    return {
      doken: {
        ...doken,
        value: RDT.make(VALUE_PARTIAL),
      },
      removed: RDT.value(doken.value),
    };
  }

  if (isSpent(doken)) {
    return {
      doken: {
        ...doken,
        id   : NONE,
        ttl  : DT.unsafeMake(0),
        value: RDT.make(VALUE_PARTIAL),
      },
    };
  }

  return {
    doken,
  };
};
