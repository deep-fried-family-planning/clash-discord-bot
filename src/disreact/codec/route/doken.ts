import {Redacted} from 'effect';
import {mutable, Number, RedactedFromSelf, type Schema, String, Struct} from 'effect/Schema';



export const DokenId        = String;
export const DokenType      = Number;
export const DokenEphemeral = Number;
export const DokenTTL       = Number;
export const DokenValue     = RedactedFromSelf(String);

const TYPE_FRESH      = -1;
const EPHEMERAL_FRESH = -1;
const PUBLIC          = 0;
const EPHEMERAL       = 1;
const VALUE_PARTIAL   = 'partial';

export const Doken = mutable(Struct({
  id       : DokenId,
  type     : DokenType,
  ephemeral: DokenEphemeral,
  ttl      : DokenTTL,
  value    : DokenValue,
}));

export type Doken = Schema.Type<typeof Doken>;

export const isFreshDoken     = (doken: Doken): boolean => doken.type === TYPE_FRESH;
export const isEphemeralDoken = (doken: Doken): boolean => doken.ephemeral === EPHEMERAL;
export const isPublicDoken    = (doken: Doken): boolean => doken.type === PUBLIC;
export const isActiveDoken    = (doken: Doken): boolean => doken.ttl > Date.now();
export const isPartialDoken   = (doken: Doken): boolean => Redacted.value(doken.value) === VALUE_PARTIAL;

export const make = (params: Doken): Doken => params;

export const makeFreshDoken = (request: any): Doken => ({
  id       : request.id,
  type     : TYPE_FRESH,
  ephemeral: EPHEMERAL_FRESH,
  ttl      : Date.now() + 2000,
  value    : Redacted.make(request.token),
});

export const makePartialDoken = (params: Omit<Doken, 'value'>): Doken => ({
  ...params,
  value: Redacted.make(VALUE_PARTIAL),
});
