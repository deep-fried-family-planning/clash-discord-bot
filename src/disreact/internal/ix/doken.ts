import {NONE_STR, Rest} from '#src/disreact/abstract/index.ts';

const TWO_SECONDS_MS      = 2 * 1000;
const FOURTEEN_MINUTES_MS = 14 * 60 * 1000;
const TWO_MINUTES_MS      = 2 * 60 * 1000;

export type DokenStatus =
  | 'active'
  | 'deferred'
  | 'deleted'
  | 'done'
  | 'expired'
  | 'none';

export type T = {
  app   : string;
  id    : string;
  token : string;
  ttl   : number;
  status: DokenStatus;
  type  : Rest.Tx;
  flags : number;
};

const initialTTL = () => Date.now() + TWO_SECONDS_MS;
const deferTTL   = () => Date.now() + FOURTEEN_MINUTES_MS;

export const makeEmpty = (): T => ({
  app   : '',
  id    : NONE_STR,
  token : NONE_STR,
  ttl   : 0,
  status: 'none',
  type  : Rest.DEFER_SOURCE,
  flags : 0,
});

export const makeFromRest = (rest: Rest.Interaction): T => ({
  app   : rest.application_id,
  id    : rest.id,
  token : rest.token,
  status: 'active',
  type  : Rest.DEFER_SOURCE,
  ttl   : initialTTL(),
  flags : 0,
});

export const makeDeferred = (doken: T): T => ({
  ...doken,
  status: 'deferred',
  ttl   : deferTTL(),
});

export const invalidateTTL = (doken: T): T => {
  if ((doken.ttl - TWO_MINUTES_MS) > Date.now()) {
    return doken;
  }
  return {
    ...doken,
    status: 'expired',
  };
};

type TDokenEncoded = {
  id   : string;
  token: string;
  ttl  : string;
  type : string;
  flags: string;
};

export const encode = (doken?: T): TDokenEncoded => doken
  ? {
    id   : doken.id,
    token: doken.token,
    ttl  : `${doken.ttl}`,
    type : `${doken.type}`,
    flags: `${doken.flags}`,
  }
  : {
    id   : NONE_STR,
    token: NONE_STR,
    ttl  : '0',
    type : '0',
    flags: '0',
  };

export const decode = (encoded: TDokenEncoded): T => invalidateTTL({
  app   : NONE_STR,
  id    : encoded.id,
  token : encoded.token,
  ttl   : parseInt(encoded.ttl),
  status: 'deferred',
  type  : parseInt(encoded.type) as unknown as Rest.Tx,
  flags : parseInt(encoded.flags),
});
