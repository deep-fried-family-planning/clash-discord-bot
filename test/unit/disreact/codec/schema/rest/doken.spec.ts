import {E, L, RDT} from '#src/internal/pure/effect.ts';
import {layer} from '@effect/vitest';
import {DateTime, Logger, LogLevel, pipe, TestClock, TestContext} from 'effect';
import * as Doken from '#src/disreact/codec/schema/rest/doken.ts';



const describe = pipe(
  L.empty,
  L.provideMerge(TestContext.TestContext),
  L.provideMerge(Logger.replace(Logger.defaultLogger, Logger.prettyLoggerDefault)),
  L.provideMerge(Logger.minimumLogLevel(LogLevel.All)),
  layer,
);



describe('Fresh Doken', (it) => {
  it.effect('when making', E.fn(function* () {
    yield* TestClock.setTime(0);

    const actual = yield* Doken.makeFresh({
      rest: {
        id            : 'testid',
        application_id: 'testappid',
        token         : 'testtoken',
      },
    });

    expect(actual).toMatchSnapshot();
  }));

  it.effect('given DateTime, when making', E.fn(function* () {
    yield* TestClock.setTime(1000);

    const actual = yield* Doken.makeFresh({
      rest: {
        id            : 'testid',
        application_id: 'testappid',
        token         : 'testtoken',
      },
      time: DateTime.unsafeMake('2025-02-21T03:51:48.193Z'),
    });

    expect(actual).toMatchSnapshot();
  }));
});



describe('Params Doken', (it) => {
  it.effect('given future TTL, when making params doken', E.fn(function* () {
    yield* TestClock.setTime(1000);

    const actual = yield* Doken.makeFromParams({
      id       : 'testid',
      ttl      : 100000000,
      app_id   : 'testappid',
      token    : RDT.make('testtoken'),
      ephemeral: 0,
      type     : 0,
    });

    expect(actual).toMatchSnapshot();
  }));

  it.effect('given past TTL, when making params doken', E.fn(function* () {
    yield* TestClock.setTime(1000);

    const actual = yield* Doken.makeFromParams({
      id       : 'testid',
      ttl      : 0,
      app_id   : 'testappid',
      token    : RDT.make('testtoken'),
      ephemeral: 0,
      type     : 0,
    });

    expect(actual).toMatchSnapshot();
  }));
});



describe('Active Doken', (it) => {
  it.effect('when activating', E.fn(function* () {
    yield* TestClock.setTime(0);

    const actual = yield* Doken.activate({
      doken: {
        _tag     : 'Doken',
        id       : 'testid',
        app_id   : 'testappid',
        token    : RDT.make('testtoken'),
        ttl      : DateTime.unsafeMake(0),
        ephemeral: 0,
        type     : 0,
        status   : 'Fresh',
      },
    });

    expect(actual).toMatchSnapshot();
  }));
});



describe('Spent Doken', (it) => {
  it.effect('when spending', E.fn(function* () {
    yield* TestClock.setTime(0);

    const actual = Doken.spend({
      doken: {
        _tag     : 'Doken',
        id       : 'testid',
        app_id   : 'testappid',
        token    : RDT.make('testtoken'),
        ttl      : DateTime.unsafeMake(0),
        ephemeral: 0,
        type     : 0,
        status   : 'Fresh',
      },
    });

    expect(actual).toMatchSnapshot();
  }));
});
