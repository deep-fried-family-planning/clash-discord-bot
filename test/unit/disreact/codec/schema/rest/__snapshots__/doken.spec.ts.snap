// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Active Doken > when activating 1`] = `
{
  "_tag": "Doken",
  "ephemeral": 0,
  "id": "testid",
  "status": "Active",
  "token": "<redacted>",
  "ttl": "1970-01-01T00:14:00.000Z",
  "type": 0,
}
`;

exports[`Fresh Doken > given DateTime, when making 1`] = `
{
  "_tag": "Doken",
  "ephemeral": 0,
  "id": "testid",
  "status": "Fresh",
  "token": "<redacted>",
  "ttl": "2025-02-21T03:51:50.693Z",
  "type": 0,
}
`;

exports[`Fresh Doken > when making 1`] = `
{
  "_tag": "Doken",
  "ephemeral": 0,
  "id": "testid",
  "status": "Fresh",
  "token": "<redacted>",
  "ttl": "1970-01-01T00:00:02.000Z",
  "type": 0,
}
`;

exports[`Params Doken > given future TTL, when making params doken 1`] = `
{
  "_tag": "Doken",
  "ephemeral": 0,
  "id": "testid",
  "status": "Active",
  "token": "<redacted>",
  "ttl": "1970-01-02T03:46:40.000Z",
  "type": 0,
}
`;

exports[`Params Doken > given past TTL, when making params doken 1`] = `
{
  "_tag": "Doken",
  "ephemeral": 0,
  "id": "testid",
  "status": "Expired",
  "token": "<redacted>",
  "ttl": "1970-01-01T00:00:00.000Z",
  "type": 0,
}
`;

exports[`Spent Doken > when spending 1`] = `
{
  "_tag": "Doken",
  "ephemeral": 0,
  "id": "testid",
  "status": "Spent",
  "token": "<redacted>",
  "ttl": "1970-01-01T00:00:00.000Z",
  "type": 0,
}
`;
